### PR TITLE
Throw an error in `bitShift*` if the shift position is out-of-bounds

### LINKED ISSUE #65516: `bitShiftLeft` may produce unexpected result.
**Describe what's wrong**
`bitShiftLeft` produces unexpected result.

**How to reproduce**

* Which ClickHouse server version to use: `24.3.3.102`
* Which interface to use, if it matters: `MySQL Shell`
* Non-default settings, if any: `distributed_product_mode='allow'`

* init database:
```sql
create database ch_main;
use ch_main;

-- sql #1
create table t_hwi130st4 ( 
  c_wpg Int32 ,
  c_s String ,
  c_i8 String ,
  c_seahk2u Float64 ,
  c_b Float64 ,
  c_w9_4ap Int32 primary key ,
  );


-- sql #3
insert into t_hwi130st4 (c_wpg, c_s, c_i8, c_seahk2u, c_b, c_w9_4ap) values 
  (-1149739257, 'oqo0b9wn0y', 'uwiw', 50.62, 1.9, -56238719), 
  (2, 'x', 'yi4n6vrr', 76.98, 9223372036854775807.8, -3), 
  (11686210, 'mlp50m', 'tc17pov2', 4294967294.0, 93.15, -1271189987), 
  (-1403535781, 'ss75f', 'zf', 3.98, 56.36, 1445090706);

-- sql #21
insert into t_hwi130st4 (c_wpg, c_s, c_i8, c_seahk2u, c_b, c_w9_4ap) values 
  (-4, 'k3ab', 'vsi9mlt2f', -0.1, 99.68, -838905824), 
  (21797751, 'yiw', 'e707iszq', 256.5, -65537.5, 1545255954), 
  (-1173571010, 'u9', 'yh_dsb95d', -4294967295.8, -4294967295.1, -86953426), 
  (1084934805, 'tod', 'a52o13ea4', 43.49, -4294967294.9, 1338049738);


-- sql #22
insert into t_hwi130st4 (c_wpg, c_s, c_i8, c_seahk2u, c_b, c_w9_4ap) values 
  (-676988243, 'g', 'fy5bz', 17.61, 96.85, -528458811), 
  (-25, 'h6mcbk8', 'pp', 97.23, -18446744073709551614.9, 5), 
  (1671090888, 'v945', 'kw', 44.4, 17.83, -1314342888), 
  (65536, 'v0zewjo49u', 'askk_1h', 72.48, 81.15, 2);


-- sql #23
alter table t_hwi130st4 add column c_skwq8 String;

-- sql #25
insert into t_hwi130st4 (c_wpg, c_s, c_i8, c_seahk2u, c_b, c_w9_4ap, c_skwq8) values 
  (-2040837324, 'si8shj', 'b', 41.20, 3.10, 1291544549, 'kx68agni4'), 
  (-310404151, '__nc8rsl', 'tqosx0s', 32767.6, 29.14, 752873682, 'v971pfad'), 
  (3, 'rdj', 'p5w', 32769.6, -9223372036854775809.1, -486995628, 'hiqsow'), 
  (-6, 'gbs780nv', 'u8ojaqgpa', 4.44, 129.0, 1746224505, 't1tzo4w538');


-- sql #26
insert into t_hwi130st4 (c_wpg, c_s, c_i8, c_seahk2u, c_b, c_w9_4ap, c_skwq8) values 
  (-1580788152, 'e', 'w6i', -9223372036854775809.2, 13.32, 962706574, 'tjl7k5r'), 
  (3, 'j', 'ix', 88.10, 89.77, -610474587, 'mh_'), 
  (1897548408, 'jb', 'ddr9kbf', 54.21, 76.36, -5, 'ditpajx'), 
  (441043873, 'wvi16h', 'r', 92.8, 30.88, 2029856608, 'pb4');


-- sql #27
insert into t_hwi130st4 (c_wpg, c_s, c_i8, c_seahk2u, c_b, c_w9_4ap, c_skwq8) values 
  (16, 'o6z', 'aujtujgdon', 30.87, 22.11, 259350534, 'ql'), 
  (1669987147, 'i4n40ete9', 'kv155wtf', 35.47, 68.80, -1976235388, 'rihep_f57'), 
  (-1000436932, 'kkpa29zk_6', 'ksckscqhr5', 9223372036854775807.9, 75.97, 1706844447, 'fsj'), 
  (950656369, 'x73x', 'hg938x95r', 70.39, 49.36, -1951482829, 'lbvo843i');

```

* Queries to run that lead to an unexpected result
```sql
SELECT
  ref_4.c_w9_4ap,
  bitShiftLeft(ref_4.c_w9_4ap, -2122704019) as c3
FROM t_hwi130st4 as ref_4
order by 1,2
;
```

**Expected behavior**
previous SQL produces:
```
mysql> SELECT
    ->   ref_4.c_w9_4ap,
    ->   bitShiftLeft(ref_4.c_w9_4ap, -2122704019) as c3
    -> FROM t_hwi130st4 as ref_4
    -> order by 1,2
    -> ;
+-------------+------+
| c_w9_4ap    | c3   |
+-------------+------+
| -1976235388 |    0 |
| -1951482829 |    0 |
| -1314342888 |    0 |
| -1271189987 |    0 |
|  -838905824 |    0 |
|  -610474587 |    0 |
|  -528458811 |    0 |
|  -486995628 |    0 |
|   -86953426 |    0 |
|   -56238719 |    0 |
|          -5 |    0 |
|          -3 |    0 |
|           2 |    0 |
|           5 |    0 |
|   259350534 |    0 |
|   752873682 |    0 |
|   962706574 |    0 |
|  1291544549 |    0 |
|  1338049738 |    0 |
|  1445090706 |    0 |
|  1545255954 |    0 |
|  1706844447 |    0 |
|  1746224505 |    0 |
|  2029856608 |    0 |
+-------------+------+
24 rows in set (0.00 sec)
Read 24 rows, 96.00 B in 0.001344 sec., 17857 rows/sec., 69.75 KiB/sec.
```

When using `bitShiftLeft` in a simple select statement:
```
mysql> select bitShiftLeft(cast(2 as Int32), -2122704019), bitShiftLeft(cast(2029856608 as Int32), -2122704019);
+---------------------------------------------+------------------------------------------------------+
| bitShiftLeft(CAST(2, 'Int32'), -2122704019) | bitShiftLeft(CAST(2029856608, 'Int32'), -2122704019) |
+---------------------------------------------+------------------------------------------------------+
|                                       16384 |                                          -1528037376 |
+---------------------------------------------+------------------------------------------------------+
1 row in set (1.80 sec)
Read 1 rows, 1.00 B in 1.802078 sec., 0 rows/sec., 0.55 B/sec.
```
The results of two queries do not match.

**Additional context**
docker compose config:
```
version: '3.8'
services:
  clickhouse-01:
    image: "clickhouse/clickhouse-server:${CHVER:-latest}"
    user: "101:101"
    container_name: clickhouse-01
    hostname: clickhouse-01
    networks:
      cluster_2S_1R:
        ipv4_address: 10.0.7.1
    volumes:
      - ${PWD}/fs/volumes/clickhouse-01/etc/clickhouse-server/config.d/config.xml:/etc/clickhouse-server/config.d/config.xml
      - ${PWD}/fs/volumes/clickhouse-01/etc/clickhouse-server/users.d/users.xml:/etc/clickhouse-server/users.d/users.xml
    depends_on:
      - clickhouse-keeper-01
      - clickhouse-keeper-02
      - clickhouse-keeper-03
  clickhouse-02:
    image: "clickhouse/clickhouse-server:${CHVER:-latest}"
    user: "101:101"
    container_name: clickhouse-02
    hostname: clickhouse-02
    networks:
      cluster_2S_1R:
        ipv4_address: 10.0.7.2
    volumes:
      - ${PWD}/fs/volumes/clickhouse-02/etc/clickhouse-server/config.d/config.xml:/etc/clickhouse-server/config.d/config.xml
      - ${PWD}/fs/volumes/clickhouse-02/etc/clickhouse-server/users.d/users.xml:/etc/clickhouse-server/users.d/users.xml
    depends_on:
      - clickhouse-keeper-01
      - clickhouse-keeper-02
      - clickhouse-keeper-03
  clickhouse-keeper-01:
    image: "clickhouse/clickhouse-keeper:${CHKVER:-latest-alpine}"
    user: "101:101"
    container_name: clickhouse-keeper-01
    hostname: clickhouse-keeper-01
    networks:
      cluster_2S_1R:
        ipv4_address: 10.0.7.5
    volumes:
      - ${PWD}/fs/volumes/clickhouse-keeper-01/etc/clickhouse-keeper/keeper_config.xml:/etc/clickhouse-keeper/keeper_config.xml

  clickhouse-keeper-02:
    image: "clickhouse/clickhouse-keeper:${CHKVER:-latest-alpine}"
    user: "101:101"
    container_name: clickhouse-keeper-02
    hostname: clickhouse-keeper-02
    networks:
      cluster_2S_1R:
        ipv4_address: 10.0.7.6
    volumes:
      - ${PWD}/fs/volumes/clickhouse-keeper-02/etc/clickhouse-keeper/keeper_config.xml:/etc/clickhouse-keeper/keeper_config.xml

  clickhouse-keeper-03:
    image: "clickhouse/clickhouse-keeper:${CHKVER:-latest-alpine}"
    user: "101:101"
    container_name: clickhouse-keeper-03
    hostname: clickhouse-keeper-03
    networks:
      cluster_2S_1R:
        ipv4_address: 10.0.7.7
    volumes:
      - ${PWD}/fs/volumes/clickhouse-keeper-03/etc/clickhouse-keeper/keeper_config.xml:/etc/clickhouse-keeper/keeper_config.xml

networks:
  cluster_2S_1R:
    driver: bridge
    ipam:
      config:
        - subnet: 10.0.7.0/24
          gateway: 10.0.7.254
```

config.xml
```xml
<clickhouse replace="true">
    <logger>
        <level>debug</level>
        <log>/var/log/clickhouse-server/clickhouse-server.log</log>
        <errorlog>/var/log/clickhouse-server/clickhouse-server.err.log</errorlog>
        <size>1000M</size>
        <count>3</count>
    </logger>
    <display_name>cluster_2S_1R node 1</display_name>
    <listen_host>0.0.0.0</listen_host>
    <http_port>8123</http_port>
    <tcp_port>9000</tcp_port>
    <mysql_port>9004</mysql_port>
    <postgresql_port>9005</postgresql_port>
    <user_directories>
        <users_xml>
            <path>users.xml</path>
        </users_xml>
        <local_directory>
            <path>/var/lib/clickhouse/access/</path>
        </local_directory>
    </user_directories>
    <distributed_ddl>
        <path>/clickhouse/task_queue/ddl</path>
    </distributed_ddl>
    <remote_servers>
        <default>
            <shard>
                <replica>
                    <host>clickhouse-01</host>
                    <port>9000</port>
                </replica>
            </shard>
            <shard>
                <replica>
                    <host>clickhouse-02</host>
                    <port>9000</port>
                </replica>
            </shard>
        </default>
    </remote_servers>
    <zookeeper>
        <node>
            <host>clickhouse-keeper-01</host>
            <port>9181</port>
        </node>
        <node>
            <host>clickhouse-keeper-02</host>
            <port>9181</port>
        </node>
        <node>
            <host>clickhouse-keeper-03</host>
            <port>9181</port>
        </node>
    </zookeeper>
    <macros>
        <shard>01</shard>
        <replica>01</replica>
    </macros>
</clickhouse>
```
users.xml
```xml
<?xml version="1.0"?>
<clickhouse replace="true">
    <profiles>
        <default>
            <max_memory_usage>10000000000</max_memory_usage>
            <use_uncompressed_cache>0</use_uncompressed_cache>
            <load_balancing>in_order</load_balancing>
            <log_queries>1</log_queries>
            <distributed_product_mode>allow</distributed_product_mode>
        </default>
    </profiles>
    <users>
        <default>
            <access_management>1</access_management>
            <profile>default</profile>
            <networks>
                <ip>::/0</ip>
            </networks>
            <password></password>
            <quota>default</quota>
            <access_management>1</access_management>
            <named_collection_control>1</named_collection_control>
            <show_named_collections>1</show_named_collections>
            <show_named_collections_secrets>1</show_named_collections_secrets>
        </default>
    </users>
    <quotas>
        <default>
            <interval>
                <duration>3600</duration>
                <queries>0</queries>
                <errors>0</errors>
                <result_rows>0</result_rows>
                <read_rows>0</read_rows>
                <execution_time>0</execution_time>
            </interval>
        </default>
    </quotas>
</clickhouse>
```

#### about us
We are the BASS team from the School of Cyber Science and Technology at Beihang University. Our main focus is on system software security, operating systems, and program analysis research, as well as the development of automated program testing frameworks for detecting software defects. Using our self-developed database vulnerability testing tool, we have identified the potential above-mentioned vulnerability that may lead to database logic error.

### PR DESCRIPTION
**Pull request by pamarcos**
Fixes #65516

Fix bitShift by throwing an error for out of bounds shift positions.
Negative values or those greater than the bit width of the value are considered out of bounds.

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Functions `bitShiftLeft` and `bitShitfRight` return an error for out of bounds shift positions (issue #65516)

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 898dd8bb8efd260733c8b967868e4fcb88fb145e with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65838/898dd8bb8efd260733c8b967868e4fcb88fb145e/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/2673a773c36f5e12b45e8c3b9b97b6cc3703afc1/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by nikitamikhaylov** [comment]
I think it should also throw an exception when we are trying to shift for more than type width. 
While this behaviour seems different from the MySQL for example, I'd better have the same behaviour as C++ compiler https://godbolt.org/z/asjGWMnbj

**Comment by pamarcos** [comment]
> I think it should also throw an exception when we are trying to shift for more than type width. While this behaviour seems different from the MySQL for example, I'd better have the same behaviour as C++ compiler https://godbolt.org/z/asjGWMnbj

Done! Even though at first I was a bit reluctant of this change because for `bitShift` I didn't see it causing any harm. However, it's true that most likely if you ever try to shift more positions than the value bit width, you're making a mistake.

File: src/Functions/bitShiftLeft.cpp [hunk 1]
```diff
@@ -24,6 +24,8 @@ struct BitShiftLeftImpl
     {
         if constexpr (is_big_int_v<B>)
             throw Exception(ErrorCodes::NOT_IMPLEMENTED, "BitShiftLeft is not implemented for big integers as second argument");
+        else if (b < 0)
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "The number of shift positions needs to be a positive value");
```
[the remark this answers is not in the document]
Comment (pamarcos) [reply src/Functions/bitShiftLeft.cpp]: Agree, in 0b24a416b5f50bb4416d4d95dddd20a1e333b569 I already changed it to `ARGUMENT_OUT_OF_BOUND`

[the remark this answers is not in the document]
Comment (pamarcos) [reply src/Functions/bitShiftLeft.cpp]: Done in 12608d2090485e5cd98f82b78c64f7014e7e391c

**Review by divanik — COMMENTED** [review]
Please, pay attention to all broken test in CI after resolving the issues and right a comment with a report, why do you think that broken tests have nothing in common with your pr

File: src/Functions/bitShiftLeft.cpp [hunk 2]
```diff
@@ -38,8 +41,12 @@ struct BitShiftLeftImpl
         else
         {
             UInt8 word_size = 8;
+            size_t n = end - pos;
+            if (b < 0 || b > B(word_size * n))
+                throw Exception(ErrorCodes::ARGUMENT_OUT_OF_BOUND, "The number of shift positions needs to be a non-negative value and less or equal to the bit width of the value to shift");
+
             /// To prevent overflow
-            if (static_cast<double>(b) >= (static_cast<double>(end - pos) * word_size) || b < 0)
+            if (static_cast<double>(b) >= (static_cast<double>(n) * word_size))
```
[the remark this answers is not in the document]
Comment (pamarcos) [reply src/Functions/bitShiftLeft.cpp]: I agree. Even though that part was there before my changes, I realized I could get rid of it while preventing overflow without using doubles. It actually made me realize a different overflow could happen in some code that I did add. The fix is in https://github.com/ClickHouse/ClickHouse/pull/65838/commits/7a7633a2309a5f1286f6120513dd75a54aefa1b6

**Comment by pamarcos** [comment]
- [ClickHouse Upgrade Check](https://s3.amazonaws.com/clickhouse-test-reports/65838/389a86ec059461f20521e4dd2c7888e2b3b37623/upgrade_check__debug_.html)
>New settings are not reflected in settings changes history (see new_settings.txt)

Unrelated

- [ClickHouse Stress Test (tsan)](https://s3.amazonaws.com/clickhouse-test-reports/65838/389a86ec059461f20521e4dd2c7888e2b3b37623/stress_test__tsan_.html)

Unrelated possible deadlock. The [test that hung](https://s3.amazonaws.com/clickhouse-test-reports/65838/389a86ec059461f20521e4dd2c7888e2b3b37623/stress_test__tsan_/hung_check.log) has 0 references to any `bitShift` function.

- [ClickHouse Stateless Tests](https://s3.amazonaws.com/clickhouse-test-reports/65838/389a86ec059461f20521e4dd2c7888e2b3b37623/stateless_tests__debug__s3_storage__[5_6].html)

Unrelated test failure, 02247_read_bools_as_numbers_json failed due to an assertion in jemalloc. [It's failed quite a few times lately](https://play.clickhouse.com/play?user=play#CldJVEgKICAgICcwMjI0N19yZWFkX2Jvb2xzX2FzX251bWJlcnNfanNvbicgQVMgbmFtZV9zdWJzdHIsCiAgICA5MCBBUyBpbnRlcnZhbF9kYXlzLAogICAgKCdTdGF0ZWxlc3MgdGVzdHMgKGFzYW4pJywgJ1N0YXRlbGVzcyB0ZXN0cyAoYWRkcmVzcyknLCAnU3RhdGVsZXNzIHRlc3RzIChhZGRyZXNzLCBhY3Rpb25zKScpIEFTIGJhY2twb3J0X2FuZF9yZWxlYXNlX3NwZWNpZmljX2NoZWNrcwpTRUxFQ1QKICAgIGNoZWNrX3N0YXJ0X3RpbWUsCiAgICBwdWxsX3JlcXVlc3RfbnVtYmVyIEFTIHByLAogICAgcmVwb3J0X3VybCBhcyByZXBvcnRfdXJsLAogICAgdGVzdF9kdXJhdGlvbl9tcywKRlJPTSBjaGVja3MKV0hFUkUgKChub3coKSAtIHRvSW50ZXJ2YWxEYXkoaW50ZXJ2YWxfZGF5cykpIDw9IGNoZWNrX3N0YXJ0X3RpbWUpIEFORCAocHVsbF9yZXF1ZXN0X251bWJlciBOT1QgSU4gKAogICAgU0VMRUNUIHB1bGxfcmVxdWVzdF9udW1iZXIgQVMgcHJuCiAgICBGUk9NIGNoZWNrcwogICAgV0hFUkUgKHBybiAhPSAwKSBBTkQgKChub3coKSAtIHRvSW50ZXJ2YWxEYXkoaW50ZXJ2YWxfZGF5cykpIDw9IGNoZWNrX3N0YXJ0X3RpbWUpIEFORCAoY2hlY2tfbmFtZSBJTiAoYmFja3BvcnRfYW5kX3JlbGVhc2Vfc3BlY2lmaWNfY2hlY2tzKSkKKSkgQU5EIChwb3NpdGlvbih0ZXN0X25hbWUsIG5hbWVfc3Vic3RyKSA+IDApIEFORCAodGVzdF9zdGF0dXMgSU4gKCdGQUlMJywgJ0VSUk9SJywgJ0ZMQUtZJykpCk9SREVSIEJZIGNoZWNrX3N0YXJ0X3RpbWUK), so I've created an issue for it: https://github.com/ClickHouse/ClickHouse/issues/66029

- [ClickHouse performance comparison](https://s3.amazonaws.com/clickhouse-test-reports/65838/389a86ec059461f20521e4dd2c7888e2b3b37623/performance_comparison__release__[4_4]/report.html)

All of the tests reported have no relation with these changes. None use `bitShift`

- [ClickHouse Integration Tests (tsan)](https://s3.amazonaws.com/clickhouse-test-reports/65838/389a86ec059461f20521e4dd2c7888e2b3b37623/integration_tests__tsan__[4_6].html)

Unrelated test error: `test_materialized_mysql_database/test.py::test_materialized_database_mysql_date_type_to_date32`

- [ClickHouse Bugfix Validation](https://s3.amazonaws.com/clickhouse-test-reports/65838/389a86ec059461f20521e4dd2c7888e2b3b37623/bugfix_validation.html)

Seems [it failed to install the deb package](https://s3.amazonaws.com/clickhouse-test-reports/65838/389a86ec059461f20521e4dd2c7888e2b3b37623/bugfix_validation/run__stateless_tests__release_.log):
```
dpkg: error: cannot access archive 'package_folder/clickhouse-common-static-dbg_*.deb': No such file or directory
```

Unrelated to the PR

- [Cloud fork sync - Stateless tests](https://d1k2gkhrlfqv31.cloudfront.net/clickhouse-test-reports-private/11816/389a86ec059461f20521e4dd2c7888e2b3b37623/stateless_tests__asan__s3_storage__[2_3].html)

Unrelated test `03037_dynamic_merges_1_vertical` failed

**Comment by divanik** [comment]
I'm sorry, my request to enumerate all the crashed tests was formulated a bit harsh. Next time just right a comment, that you scanned all the tests and all of them seems unrelated to your PR or already flaky.

 It would be great if you create issues about some unreported crashes from your CI run. Usually these crashes are already reported, but sometimes some new bugs can be found

**Comment by pamarcos** [comment]
> I'm sorry, my request to enumerate all the crashed tests was formulated a bit harsh. Next time just write a comment, that you scanned all the crashed jobs and all of them seems unrelated to your PR or are already flaky

Ok, no worries. Understood.

> It would be great if you create issues about some unreported crashes from your CI run. Usually these crashes are already reported, but sometimes some new bugs can be found

I had already created an issue that it was clearly a flaky test. I've reviewed again the rest and created issues for those missing. I edited [my comment](https://github.com/ClickHouse/ClickHouse/pull/65838#issuecomment-2205570160) with the new issues created.

Let me know if there's anything else missing in the PR that I need to address.

**Comment by divanik** [comment]
> Let me know if there's anything else missing in the PR that I need to address.

Could you take a look at comment about the right part of the expression const UInt256 bit_limit = word_size * n?

**Comment by pamarcos** [comment]
> Could you take a look at comment about the right part of the expression const UInt256 bit_limit = word_size * n?

@divanik I don't see any comment anywhere :cry: . Are you sure you submitted the review/comment? If so, please share a link to the comment, because there's nothing I can see as of now.

File: src/Functions/bitShiftLeft.cpp [hunk 3]
```diff
@@ -37,9 +40,12 @@ struct BitShiftLeftImpl
             throw Exception(ErrorCodes::NOT_IMPLEMENTED, "BitShiftLeft is not implemented for big integers as second argument");
         else
         {
-            UInt8 word_size = 8;
-            /// To prevent overflow
-            if (static_cast<double>(b) >= (static_cast<double>(end - pos) * word_size) || b < 0)
+            const UInt8 word_size = 8 * sizeof(*pos);
+            size_t n = end - pos;
+            const UInt256 bit_limit = word_size * n;
```
[the remark this answers is not in the document]
Comment (pamarcos) [reply src/Functions/bitShiftLeft.cpp]: Fixed in https://github.com/ClickHouse/ClickHouse/pull/65838/commits/e7e2b0953c0df57c95f6b5e0f6cc1afe5914c4e9
Just realized I wrote `buffer overflow` in the commit message when I meant just `overflow` :sweat:

**Review by divanik — APPROVED** [review]

**Comment by pamarcos** [comment]
Two more issues created after checking that the some CI check failed: https://github.com/ClickHouse/ClickHouse/issues/66116 and https://github.com/ClickHouse/ClickHouse/issues/66117

**Comment by UnamedRus** [comment]
> However, it's true that most likely if you ever try to shift more positions than the value bit width, you're making a mistake.

It's "wrong"

CH always try to use least type possible, so in case of literals, you will get different behavior based on typed value itself.
And this behavior is somewhat special, other DBMS usually use much wider type like 64 bit by default

```
SELECT bitShiftLeft(256, 10);

0

SELECT bitShiftLeft(1, 10)

Received exception from server (version 24.11.1):
Code: 69. DB::Exception: Received from localhost:9000. DB::Exception: The number of shift positions needs to be a non-negative value and less or equal to the bit width of the value to shift: In scope SELECT bitShiftLeft(1, 10). (ARGUMENT_OUT_OF_BOUND)
(query: SELECT bitShiftLeft(1, 10))
```

https://fiddle.clickhouse.com/6063d839-c924-4e88-a3f7-0022ff803ed0

**Comment by UnamedRus** [comment]
> As a workaround in ClickHouse you can always cast to whatever you want to bypass this exception:

So, basically, it does mean that you "always" need to do additional cast within bitShift operator, because there are functions exist in CH, which return smaller types than arguments. (modulo as example).

**Pull request merged**